File: frontend/src/types/github.ts
export type GitHubItem = {
  id: string;
  number: number;
  title: string;
  state: string;
  summary?: string;
  html_url: string;
  created_at: string;
  updated_at: string;
  comments: number;
  labels: string[];
  user: {
    login: string;
    avatar_url: string;
    profile_url: string;
  };
  reactions?: number;
  is_pull_request: boolean;
  assignees?: {
    login: string;
    avatar_url: string;
    profile_url: string;
  }[];
  merged?: boolean;
};

export type Contributor = {
  username: string;
  avatar_url: string;
  profile_url: string;
  tldr: string;
  prs: GitHubItem[];
  issues: GitHubItem[];
};

export type PeopleData = {
  username: string;
  avatar_url: string;
  profile_url: string;
  tldr: string;
  prs: GitHubItem[];
  issues: GitHubItem[];
};

export type Timeframe = "last_day" | "last_week" | "last_month" | "last_year";

export type TLDRData = {
  prs: GitHubItem[] | null;
  issues: GitHubItem[] | null;
  people: PeopleData[] | null;
  tldr: string | null;
};

export type StoredTLDRReport = {
  id: string; // `${repo}:${timeframe}`
  repo: string;
  timeframe: Timeframe;
  data: TLDRData;
  generatedAt: string; // ISO timestamp
  version: number; // For future migration
};

export type RepoSummary = {
  name: string;
  full_name: string;
  description: string | null;
  html_url: string;
  private: boolean;
  fork: boolean;
  archived: boolean;
  language: string | null;
  stargazers_count: number;
  updated_at: string;
};

File: backend/models/github.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class GithubUser(BaseModel):
    login: str
    id: int
    avatar_url: str
    html_url: str


class GitHubItem(BaseModel):
    id: int
    number: int
    title: str
    body: str
    summary: Optional[str] = None
    user: GithubUser
    html_url: str
    state: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    comments: int
    reactions: int
    labels: list[str]
    is_pull_request: bool
    merged: Optional[bool] = False
    assignees: Optional[list[GithubUser]] = None
    author_association: Optional[str] = None


class ContributorActivity(BaseModel):
    username: str
    avatar_url: Optional[str]
    profile_url: Optional[str]
    tldr: Optional[str]
    prs: list[GitHubItem]
    issues: list[GitHubItem]


class PatchItem(BaseModel):
    file: str
    patch: str
